File: sites/newEgg.js
const delay = require('delay');
const cheerio = require('cheerio');
const axios = require('axios').default;
const fs = require('fs');
const { gzip } = require('zlib');
const { response } = require('express');
const { resolve } = require('path');
const { json } = require('body-parser');
const fetch = require('node-fetch');
const Discord = require('discord.js');
// require ('newrelic');
const rp = require('request-promise').defaults({
	followAllRedirects: true,
	resolveWithFullResponse: true,
	gzip : true,
    timeout: 5000

});
const { newEggWebhook, companyImage , companyColor} = require('../config.json')
let Tokens = newEggWebhook.split('https://discordapp.com/api/webhooks/')[1]
let channelId = Tokens.split('/')[0]
let discordToken = Tokens.split('/')[1]
// https://discordapp.com/api/webhooks/816740348222767155/2APr1EdhzNO4hRWznexhMRlO0g7qOiCkI7HFtmuU7_r48PCWnGYmSTGJmRVX0LPCNN_t
const webhookClient1 = new Discord.WebhookClient(channelId, discordToken);

class newEggMonitor {
    constructor(sku) {
        this.trueSku = sku
        this.sku = ''
        this.skuName = sku
        this.delay = 850000; // this.delay = 390000
        this.startDelay = 0; //  this.startDelay = 6000;
        this.availability = '';
        this.stockNumber = '';
        this.proxyList = [];
        this.isStock = false
        this.imageUrl = ''
        this.productName = ''

    }

    async task () {
        try {
            console.log('Start')
            console.log(this.sku)
            console.log(this.skuName)
            await this.getProxies()
         //   console.log(this.proxyList)
            await this.monitor()
        } catch (error) {
            fs.appendFileSync('./errors.txt', error.toString() + '\n', (err =>{
                console.log(err)
            }))
        }
    }
    async getProxies() {
		try {
			// read contents of the file
			const data = fs.readFileSync('proxies.txt', 'utf-8');

			const lines = data.split(/\r?\n/);

			// print all lines
			lines.forEach((line) => {
				const lineSplit = line.split(':');
				const item1 = {
					ip : lineSplit[0],
					port: lineSplit[1],
					userAuth: lineSplit[2],
					userPass: lineSplit[3],
				};
				this.proxyList.push(item1);

				// console.log(line);
				// console.log(item1);
				// console.log('\n\n\n\n\n');
			});
		} catch (err) {
			console.error(err);
            fs.appendFileSync('./errors.txt', error.toString() + '\n', (err =>{
                console.log(err)
            }))
        }
    }
    async monitor () {
       try {
        console.log('Starting Monitoring')
        var testing = ''
        return new Promise( async ( resolve, reject) => {

                let i = 0
                var { skuBank } = require('../dms')
                let index = skuBank.findIndex(e => e.sku == this.trueSku)

               try {
                   console.log(this.skuName)
                    let fetchProductPage = await rp.get({
                    url : `https://www.newegg.com/prada/p/${this.skuName}`,
                    headers : {
                    "accept": "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9",
                    "accept-language": "en-US,en;q=0.9",
                    "cache-control": "no-cache",
                    "pragma": "no-cache",
                    "sec-fetch-dest": "document",
                    "sec-fetch-mode": "navigate",
                    "sec-fetch-site": "none",
                    "sec-fetch-user": "?1",
                    "upgrade-insecure-requests": "1"
                    }
                })
                this.sku = fetchProductPage?.body?.split('/ProductImage/')[1].split('-')[0] + '-'
                this.sku = this.sku + fetchProductPage?.body?.split('/ProductImage/')[1].split('-')[1] + '-'
                this.sku = this.sku + fetchProductPage?.body?.split('/ProductImage/')[1].split('-')[2] 
                console.log(this.sku)
                } catch (error) {
                    console.log(error.message)
                    skuBank[index].name = 'Restart'
                    skuBank[index]["error"] = error.message
                }

                    while(!skuBank[index]?.stop){
                        skuBank[index].name = this.productName
                        if(i+1 == this.proxyList.length){
                            i = 0
                        }
                        let proxy = this.proxyList[i]
                        i++
                        console.log(`${proxy.userAuth}:${proxy.userPass}@${proxy.ip}:${proxy.port}`)
                        try {
                            let fetchSite = await rp.get({
                                url : `https://www.newegg.com/product/api/ProductRealtime?ItemNumber=${this.sku}`,
                                proxy : `http://${proxy.userAuth}:${proxy.userPass}@${proxy.ip}:${proxy.port}`
                            })   
                           // console.log(fetchSite.body)
                           let parsedBod; 
                           let image;
                            let originalPrice; 
                            let currentPrice;
                           if(fetchSite?.body?.includes("We apologize for the confusion, but we can't quite tell if you're a person or?")){
                            //   await delay(100000)
                            console.log('Captcha')
                           } else {
                            testing = fetchSite.body
                            let parsedBod = JSON.parse(fetchSite?.body)
                            image = parsedBod?.MainItem?.Image?.ItemCellImageName
                            
                            this.productName = parsedBod?.MainItem?.Description?.Title
                            originalPrice = parsedBod?.MainItem?.OriginalUnitPrice
                            currentPrice = parsedBod?.MainItem?.FinalPrice
                            this.availability = parsedBod?.MainItem?.Instock
                            this.stockNumber = parsedBod?.MainItem?.Stock 
                            console.log(`Task ${i} : ${fetchSite.statusCode}`, this.availability, this.stockNumber, this.productName, this.isStock)
                           }
                            if(!this.isStock && this.availability) {
                                // Send in stock webhook
                                console.log(`Task ${i} : ${this.isStock} and ${this.availability}`)
                                this.isStock = true
                                let embed1 = new Discord.MessageEmbed()
                                .setColor(companyColor)
                                .setTitle('New Egg Monitor')
                                .setURL(`https://www.newegg.com/Prada/p/${this.skuName}`)
                                .addFields(
                                    { name : 'Product Name', value : `${this.productName}`},
                                    { name : 'Product Availability', value : `Product In Stock`, inline : true},
                                    { name : 'Stock Number', value : `${this.stockNumber}`, inline : true}, 
                                    { name : 'Current Price', value : currentPrice, inline : true} 
                                    )
                                .addField("Links", `[ATC](https://secure.newegg.com/shopping/addtocart.aspx?submit=add&itemList=${this.sku}) | [Cart](https://secure.newegg.com/shop/cart)`)
                                // .addField('Original Price', originalPrice, true)
                                // .addField('Current Price', currentPrice, true)
                                .setThumbnail(`http://c1.neweggimages.com/ProductImageOriginal/${image}`)
                                .setTimestamp()
                                .setFooter('Prada#4873', 'https://cdn.discordapp.com/attachments/772173046235529256/795132477659152444/pradakicks.jpg');
                                webhookClient1.send({
                                    username: 'New Egg',
                                    avatarURL: companyImage,
                                    embeds: [embed1],
                                })
                            } else if (!this.availability && this.isStock) {
                                this.isStock = false
                            } else if (!this.availability && !this.isStock){
                                this.isStock = false
                                this.availability = false
                                // Not Important
                              //  console.log('False false')
                            } else if(this.availability && this.isStock){
                                this.availability = true
                                this.isStock = true
                                // Not Important
                                // console.log(true, true)
                            } else {
                                fs.appendFileSync('what.txt', this.availability + this.isStock + '\n', (err =>{
                                    console.log(err)
                                }))
                            }
        
                        
                            // console.log(originalPrice, currentPrice)
                           // console.log(parsedBod)
            
            
                        } catch (error) {
                            console.log(error)
                            // fs.appendFileSync('./errors.txt', error.toString() + '\n', (err =>{
                            //     console.log(err)
                            // }))
                            
                              if(error.message.includes('Unexpected token')){
                                  if(testing.includes('<!DOCTYPE html><html')){
                                    console.log("HTML")
                                  } else if (testing.includes('Are you a human?')) {
                                      console.log('Captcha')
                                  } else {
                                        console.log('No HTML')
                                        console.log(testing)
                                  }
                               
                            } else if (error.message.includes('403')){
                                // await delay(400000)
                                console.log('403 Access Denied')
                            }
                        }
                        await delay(this.startDelay)
                    }
                    console.log('stopped!')
                    resolve('Stopped')
                    return
              
        })
       } catch (error) {
        fs.appendFileSync('bigError.txt', error.toString() + '\n', (err =>{
            console.log(err)
        }))
       }
  
    }

}


module.exports = {
    newEggMonitor
}

File: GoMonitor/sites/newEgg/newEgg.go
package NewEggMonitor

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"net/http"
	"net/url"
	"os"
	"strings"
	"time"

	"github.com/bradhe/stopwatch"
	FetchProxies "github.con/prada-monitors-go/helpers/proxy"
)

type Config struct {
	sku              string
	skuName          string // Only for new Egg
	startDelay       int
	discord          string
	site             string
	priceRangeMax    int
	priceRangeMin    int
	proxyCount       int
	indexMonitorJson int
}
type Monitor struct {
	Config              Config
	monitorProduct      Product
	Availability        bool
	currentAvailability string
	Client              http.Client
	file                *os.File
	stop                bool
	CurrentCompanies    []Company
}
type Product struct {
	name        string
	stockNumber int
	offerId     string
	price       int
	image       string
}
type Proxy struct {
	ip       string
	port     string
	userAuth string
	userPass string
}
type ItemInMonitorJson struct {
	Sku       string `json:"sku"`
	Site      string `json:"site"`
	Stop      bool   `json:"stop"`
	Name      string `json:"name"`
	Companies []Company
}
type Company struct {
	Company      string `json:"company"`
	Webhook      string `json:"webhook"`
	Color        string `json:"color"`
	CompanyImage string `json:"companyImage"`
}
type NewEggResponse struct {
	MainItem struct {
		Description struct {
			ShortTitle string `json:"ShortTitle"`
			Title      string `json:"Title"`
		} `json:"Description"`
		FinalPrice float64 `json:"FinalPrice"`
		Image      struct {
			ImagePathPattern []struct {
				PathPattern string `json:"PathPattern"`
				Size        int64  `json:"Size"`
			} `json:"ImagePathPattern"`
			ImageType         int64  `json:"ImageType"`
			IsDFISImage       bool   `json:"IsDFISImage"`
			ItemCellImageName string `json:"ItemCellImageName"`
			Normal            struct {
				ImageName     string `json:"ImageName"`
				ImageNameList string `json:"ImageNameList"`
			} `json:"Normal"`
		} `json:"Image"`
		Instock  bool `json:"Instock"`
		NewImage struct {
			DFIS360ImgFlag interface{} `json:"DFIS360ImgFlag"`
			ImageName      string      `json:"ImageName"`
			ImageNameList  string      `json:"ImageNameList"`
		} `json:"NewImage"`
		Stock         int64 `json:"Stock"`
		StockForCombo int64 `json:"StockForCombo"`
	} `json:"MainItem"`
}

func NewMonitor(sku string, skuName string, priceRangeMin int, priceRangeMax int) *Monitor {
	defer func() {
		if r := recover(); r != nil {
			fmt.Printf("Site : %s, Product : %s Recovering from panic in printAllOperations error is: %v \n", sku, sku, r)
		}
	}()
	fmt.Println("TESTING", sku, skuName, priceRangeMin, priceRangeMax)
	m := Monitor{}
	m.Availability = false
	// var err error
	m.Client = http.Client{Timeout: 10 * time.Second}
	m.Config.site = "NewEgg"
	m.Config.startDelay = 3000
	m.Config.sku = sku
	m.Config.skuName = skuName
	// 	m.file, err = os.Create("./testing.txt")
	m.Client = http.Client{Timeout: 60 * time.Second}
	m.Config.discord = "https://discord.com/api/webhooks/816740348222767155/2APr1EdhzNO4hRWznexhMRlO0g7qOiCkI7HFtmuU7_r48PCWnGYmSTGJmRVX0LPCNN_t"
	m.monitorProduct.name = "Testing Product"
	m.monitorProduct.stockNumber = 10
	m.Config.priceRangeMax = priceRangeMax
	m.Config.priceRangeMin = priceRangeMin

	proxyList := FetchProxies.Get()

	// fmt.Println(timeout)
	//m.Availability = "OUT_OF_STOCK"
	//fmt.Println(m)
	go m.checkStop()
	time.Sleep(3000 * (time.Millisecond))
	i := true
	for i == true {
		if !m.stop {
			currentProxy := m.getProxy(proxyList)
			splittedProxy := strings.Split(currentProxy, ":")
			proxy := Proxy{splittedProxy[0], splittedProxy[1], splittedProxy[2], splittedProxy[3]}
			//	fmt.Println(proxy, proxy.ip)
			prox1y := fmt.Sprintf("http://%s:%s@%s:%s", proxy.userAuth, proxy.userPass, proxy.ip, proxy.port)
			proxyUrl, err := url.Parse(prox1y)
			if err != nil {
				fmt.Println(err)

				return nil
			}
			defaultTransport := &http.Transport{
				Proxy: http.ProxyURL(proxyUrl),
			}
			m.Client.Transport = defaultTransport
			go m.monitor()
			time.Sleep(250 * (time.Millisecond))
		} else {
			fmt.Println(m.Config.skuName, "STOPPED STOPPED STOPPED")
			i = false
		}

	}
	return &m
}

func (m *Monitor) monitor() error {
	watch := stopwatch.Start()
	defer func() {
		if r := recover(); r != nil {
			fmt.Printf("Site : %s, Product : %s Recovering from panic in printAllOperations error is: %v \n", m.Config.site, m.Config.skuName, r)
		}
	}()
	url := fmt.Sprintf("https://www.newegg.com/product/api/ProductRealtime?ItemNumber=%s", m.Config.skuName)
	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		fmt.Println(err)
		return nil
	}
	req.Header.Add("pragma", "no-cache")
	req.Header.Add("cache-control", "no-cache")
	req.Header.Add("accept", "application/json")
	req.Header.Add("dnt", "1")
	req.Header.Add("accept-language", "en")
	req.Header.Add("user-agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/87.0.4280.141 Safari/537.36")
	req.Header.Add("content-type", "application/json")
	req.Header.Add("sec-fetch-site", "cross-site")
	req.Header.Add("sec-fetch-mode", "cors")
	req.Header.Add("sec-fetch-dest", "empty")
	req.Header.Set("Connection", "close")
	req.Close = true
	res, err := m.Client.Do(req)
	if err != nil {
		fmt.Println(err)
		return nil
	}
	var monitorAvailability bool
	defer res.Body.Close()
	defer func() {
		watch.Stop()
		fmt.Printf("New Egg Monitor : %s, %d, %t, %t, Status Code : %d, %d Milliseconds elapsed: %v \n", m.monitorProduct.name, m.monitorProduct.price, m.Availability, monitorAvailability, m.monitorProduct.stockNumber, res.StatusCode, watch.Milliseconds())
	}()
	body, err := ioutil.ReadAll(res.Body)
	if err != nil {
		fmt.Println(err)
		return nil
	}
	if res.StatusCode != 200 {
		return nil
	}
	var realBody NewEggResponse
	err = json.Unmarshal(body, &realBody)
	if err != nil {
		fmt.Println(err)
		return nil
	}
	m.monitorProduct.name = realBody.MainItem.Description.Title
	m.monitorProduct.price = int(realBody.MainItem.FinalPrice)
	monitorAvailability = realBody.MainItem.Instock
	m.monitorProduct.stockNumber = int(realBody.MainItem.Stock)
	m.monitorProduct.image = realBody.MainItem.Image.ItemCellImageName

	if m.Availability == false && monitorAvailability == true {
		fmt.Println("Item in Stock")
		go m.sendWebhook()
	}
	if m.Availability == true && monitorAvailability == false {
		fmt.Println("Item Out Of Stock")
	}
	m.Availability = monitorAvailability
	return nil
}

func (m *Monitor) getProxy(proxyList []string) string {
	if m.Config.proxyCount+1 == len(proxyList) {
		m.Config.proxyCount = 0
	}
	m.Config.proxyCount++
	return proxyList[m.Config.proxyCount]
}

func (m *Monitor) sendWebhook() error {
	defer func() {
		if r := recover(); r != nil {
			fmt.Printf("Site : %s, Product : %s Recovering from panic in printAllOperations error is: %v \n", m.Config.site, m.Config.skuName, r)
		}
	}()
	for _, letter := range m.monitorProduct.name {
		if string(letter) == `"` {
			m.monitorProduct.name = strings.Replace(m.monitorProduct.name, `"`, "", -1)
		}
	}
	for _, comp := range m.CurrentCompanies {
		fmt.Println(comp.Company)
		go webHookSend(comp, m.Config.site, m.Config.skuName, m.monitorProduct.name, m.monitorProduct.price, m.monitorProduct.stockNumber, "test", m.monitorProduct.image)
	}
	return nil
}
func webHookSend(c Company, site string, sku string, name string, price int, stockNumber int, time string, image string) {
	payload := strings.NewReader(fmt.Sprintf(`{
		"content": null,
		"embeds": [
		  {
			"title": "%s Monitor",
			"url": "https://www.newegg.com/Prada/p/%s",
			"color": %s,
			"fields": [
			  {
				"name": "Product Name",
				"value": "%s"
			  },
			  {
				"name": "Product Availability",
				"value": "In Stock",
				"inline": true
			  },
			  {
				"name": "Price",
				"value": "%d",
				"inline": true
			  },
			  {
				"name": "Sku",
				"value": "%s",
				"inline": true
			  },
			  {
				"name": "Stock Number",
				"value": "%d"
			  },
			  {
				"name": "Links",
				"value": "[ATC](https://secure.newegg.com/shopping/addtocart.aspx?submit=add&itemList=%s) | [Cart](https://secure.newegg.com/shop/cart)"
			  }
			],
			"footer": {
			  "text": "Prada#4873"
			},
			"timestamp": "2021-05-13 13:57:26.5157268",
			"thumbnail": {
			  "url": "http://c1.neweggimages.com/ProductImageOriginal/%s"
			}
		  }
		],
		"avatar_url": "%s"
	  }`, site, sku, c.Color, name, price, sku, stockNumber, sku, image, c.CompanyImage))
	req, err := http.NewRequest("POST", c.Webhook, payload)
	if err != nil {
		fmt.Println(err)
	}
	req.Header.Add("pragma", "no-cache")
	req.Header.Add("cache-control", "no-cache")
	req.Header.Add("accept", "application/json")
	req.Header.Add("dnt", "1")
	req.Header.Add("accept-language", "en")
	req.Header.Add("user-agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/87.0.4280.141 Safari/537.36")
	req.Header.Add("content-type", "application/json")
	req.Header.Add("sec-fetch-site", "cross-site")
	req.Header.Add("sec-fetch-mode", "cors")
	req.Header.Add("sec-fetch-dest", "empty")
	res, err := http.DefaultClient.Do(req)
	if err != nil {
		fmt.Println(err)
	}
	defer res.Body.Close()
	fmt.Println(res)
	fmt.Println(payload)
	return
}
func (m *Monitor) checkStop() error {

	for !m.stop {
		defer func() {
			if r := recover(); r != nil {
				fmt.Printf("Site : %s, Product : %s  CHECK STOP Recovering from panic in printAllOperations error is: %v \n", m.Config.site, m.Config.sku, r)
			}
		}()
		getDBPayload := strings.NewReader(fmt.Sprintf(`{
			"site" : "%s",
			"sku" : "%s"
		  }`, strings.ToUpper(m.Config.site), m.Config.sku))
		url := fmt.Sprintf("http://localhost:7243/DB")
		req, err := http.NewRequest("POST", url, getDBPayload)
		if err != nil {
			fmt.Println(err)
			return nil
		}
		res, err := http.DefaultClient.Do(req)
		if err != nil {
			fmt.Println(err)
			res.Body.Close()
			return nil

		}
		body, err := ioutil.ReadAll(res.Body)
		if err != nil {
			fmt.Println(err)
			res.Body.Close()
			return nil
		}
		var currentObject ItemInMonitorJson
		err = json.Unmarshal(body, &currentObject)
		if err != nil {
			fmt.Println(err)
			res.Body.Close()
			return nil
		}
		m.stop = currentObject.Stop
		m.CurrentCompanies = currentObject.Companies
		fmt.Println(m.CurrentCompanies)
		res.Body.Close()
		time.Sleep(5000 * (time.Millisecond))
	}
	return nil
}
